### PR TITLE
Restructure into python package

### PR DESCRIPTION
**Pull request by Mukundan314**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    long_description = fd.read()
+
+setup(
+    name="PyRival",
+    use_scm_version={
+        "version_scheme": "post-release",
+        "write_to": "pyrival/version.py"
+    },
+    url="https://github.com/cheran-senthil/PyRival",
+    author="Cheran Senthil",
```
Comment (cheran-senthil) [inline setup.py]: Mention all authors

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    author_email="hello@cheran.io",
+    classifiers=[
+        "Development Status :: 4 - Beta",
+        "License :: OSI Approved :: MIT License",
```
Comment (cheran-senthil) [inline setup.py]: We are no longer using MIT

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        "Programming Language :: Python",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.4",
+        "Programming Language :: Python :: 3.5",
+        "Programming Language :: Python :: 3.6",
+        "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3 :: Only",
+        "Programming Language :: Python :: Implementation :: CPython",
+        "Programming Language :: Python :: Implementation :: PyPy",
```
Comment (cheran-senthil) [inline setup.py]: Will need to add fixes for Python 2

**Comment by cheran-senthil** [comment]
Perhaps we can add namespace for some of the functions like fft, ntt instead of directly calling.

**Comment by Mukundan314** [comment]
## Problems

#### Problem

cumsum2d uses global vars

#### Possible Solutions

- Convert it into a class

---

#### Problem

nCr_p does pre-computations based on mod this makes changing the mod not possible

#### Possible Solutions

- Convert it into a class
- Add function to change the mod

---

#### Problem

Hashing class uses global mod.

#### Possible Solutions

- Store mod inside class

---

#### Problem

Modular import is not possible

---

**Pull request merged**